crates/abi/src/function.rs

```diff
@@ -14,6 +14,36 @@ pub enum StateMutability {
     Payable,
 }
 
+impl StateMutability {
+    pub fn decision(mut_self: Option<bool>, mut_ctx: Option<bool>) -> Self {
+        // Check ABI conformity
+        // https://github.com/ethereum/fe/issues/558
+        //
+        //             |none self             |    self  |  mut self             |
+        // | none ctx  |    pure              |    view  |  payable/nonpayable   |
+        // | ctx       |    view              |    view  |  payable/nonpayable   |
+        // | mut ctx   |   payable/nopayable  |    view  |  payable/nonpayable   |
```

Comment:
@cburgdorf `mut ctx` + `self` should be `payable/non payable`, right?

Reply:
Yes, as soon as we have `mut` in a public contract function, then it needs to be `payable`

Reply:
I think more sense if we only allow `mut ctx` follow after `mut self` because we want to `modify` something.

Reply:
These are two different things. You can take `self` non mutable but take `mut ctx` because e.g. you want to emit an event. So you aren't changing any storage of the contract but you are still making changes to the blockchain. In terms of the ABI both of these should be marked as `payable`. There's a longer form write-up of the different access classes under that scheme: https://github.com/ethereum/fe/issues/558